README.md
Mediacore Cookbook
==================
This cookbook installs and configures a mediacore server. 

Requirements
------------

#### cookbooks 
- `python` - https://github.com/opscode-cookbooks/python 
- `nginx` - https://github.com/opscode-cookbooks/nginx 
- `supervisor` - https://github.com/opscode-cookbooks/supervisor 
- `postgresql` - https://github.com/opscode-cookbooks/postgresql  

Attributes
----------

#### mediacore::default
<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['mediacore']['user']</tt></td>
    <td>String</td>
    <td>mediacore user</td>
    <td><tt>www-data</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['group']</tt></td>
    <td>String</td>
    <td>mediacore group</td>
    <td><tt>www-data</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['smtp_server']</tt></td>
    <td>String</td>
    <td>smtp server for mediacore to use</td>
    <td><tt>localhost</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['error_email_from']</tt></td>
    <td>String</td>
    <td>email mediacore uses for errors</td>
    <td><tt>mediacore@localhost</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['email_to']</tt></td>
    <td>String</td>
    <td>email mediacore sends errors to</td>
    <td><tt>you@yourdomain.com</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['host']</tt></td>
    <td>String</td>
    <td>host mediacore binds to</td>
    <td><tt>0.0.0.0</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['port']</tt></td>
    <td>String</td>
    <td>port mediacore binds to</td>
    <td><tt>8080</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['db_type']</tt></td>
    <td>String</td>
    <td>dbms to use for mediacore</td>
    <td><tt>postgresql</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['db_user']</tt></td>
    <td>String</td>
    <td>mediacore database user</td>
    <td><tt>mediacore</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['db_pass']</tt></td>
    <td>String</td>
    <td>mediacore database password</td>
    <td><tt>mediacore</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['db_address']</tt></td>
    <td>String</td>
    <td>database location</td>
    <td><tt>localhost</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['database']</tt></td>
    <td>String</td>
    <td>mediacore database</td>
    <td><tt>mediacore</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['sqlalchemy']['url']</tt></td>
    <td>String</td>
    <td>mediacore sqlalchemy url</td>
    <td><tt>postgresql://mediacore:mediacore@localhost/mediacore</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['sqlalchemy']['echo']</tt></td>
    <td>String</td>
    <td>sqlalchemy loging</td>
    <td><tt>False</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['sqlalchemy']['pool_recycle']</tt></td>
    <td>String</td>
    <td>sqlalchemy pool recycling time</td>
    <td><tt>3600</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['version']</tt></td>
    <td>String</td>
    <td>mediacore version to use</td>
    <td><tt>0.10.0</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['dir']</tt></td>
    <td>String</td>
    <td>where to install mediacore</td>
    <td><tt>/usr/local/mediacore</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['data_storage_dir']</tt></td>
    <td>String</td>
    <td>where to store data files such as uploads</td>
    <td><tt>/usr/local/mediacore/data</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['venv']</tt></td>
    <td>String</td>
    <td>mediacore virtualenv location</td>
    <td><tt>/usr/local/mediacore/venv</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['git_repo']</tt></td>
    <td>String</td>
    <td>mediacore git repo</td>
    <td><tt>git://github.com/mediacore/mediacore-community.git</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['log_location']</tt></td>
    <td>String</td>
    <td>mediacore log location</td>
    <td><tt>/var/log/mediacore</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['plugins']</tt></td>
    <td>String</td>
    <td>list of plugins for config file</td>
    <td><tt>*</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['uwsgi']['socket']</tt></td>
    <td>String</td>
    <td>mediacore socket for uwsgi</td>
    <td><tt>/tmp/uwsgi-mediacore.sock</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['uwsgi']['master']</tt></td>
    <td>String</td>
    <td>uwsgi is master</td>
    <td><tt>true</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['uwsgi']['processes']</tt></td>
    <td>String</td>
    <td>number of uwsgi processes</td>
    <td><tt>5</tt></td>
  </tr>
</table>

#### mediacore::web_server
<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['mediacore']['web_server']['type']</tt></td>
    <td>String</td>
    <td>web server to use</td>
    <td><tt>nginx</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['web_server']['port']</tt></td>
    <td>String</td>
    <td>port to bind webserver to</td>
    <td><tt>80</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['web_server']['server_name']</tt></td>
    <td>String</td>
    <td>server_name to use for web server</td>
    <td><tt>node['hostname']</tt></td>
  </tr>
  <tr>
    <td><tt>['mediacore']['web_server']['client_max_body_size']</tt></td>
    <td>String</td>
    <td>max client body size</td>
    <td><tt>1500M</tt></td>
  </tr>
</table>

Usage
-----
#### mediacore::default

Just include `mediacore` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[mediacore]"
  ]
}
```

Contributing
------------
1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write you change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Authors:

* Jim Rosser - jarosser06@arch.tamu.edu 


CHANGELOG.md
# CHANGELOG for mediacore

## 0.2.1: 

* cookbook actually works now from scratch

## 0.2.0:

* Added database recipe with support for postgres
* various bug fixes, typos
* using uwsgi paster

## 0.1.0:

* Initial release of mediacore


recipes/deploy.rb
include_recipe "mediacore::packages"
include_recipe "mediacore::web_server"

[ node[:mediacore][:dir],
  node[:mediacore][:log_location]
].each do |dir|
  directory dir do
    action :create
    owner node[:mediacore][:user]
    group node[:mediacore][:group]
    mode "0775"
  end
end

git node[:mediacore][:dir] do
  repository node[:mediacore][:git_repo]
  reference "v#{node[:mediacore][:version]}"
  user node[:mediacore][:user]
  group node[:mediacore][:group]
  notifies :run,"execute[mediacore_setup]",:delayed
end

directory node[:mediacore][:venv] do
  action :create
  owner node[:mediacore][:user]
  group node[:mediacore][:group]
  mode "0775"
end

python_virtualenv node[:mediacore][:venv] do
  owner node[:mediacore][:user]
  group node[:mediacore][:group]
  options "--distribute --no-site-packages"
  action :create
end

template "#{node[:mediacore][:dir]}/deployment.ini" do
  action :create
  mode "0755"
  owner node[:mediacore][:user]
  group node[:mediacore][:group]
  source "deployment.ini.erb"
end

python_pip "uwsgi" do
  action :install
end

case node[:mediacore][:db_type]
when "postgresql"
  include_recipe "postgresql::client"
  python_pip "psycopg2" do
    action :install
    virtualenv node[:mediacore][:venv]
  end
end
  
execute "mediacore_setup" do
  user node[:mediacore][:user]
  cwd node[:mediacore][:dir]
  command "#{node[:mediacore][:venv]}/bin/python setup.py develop"
  action :nothing
  notifies :run, "execute[application_migration]", :delayed
end

execute "application_migration" do
  user node[:mediacore][:user]
  cwd node[:mediacore][:dir]
  command "#{node[:mediacore][:venv]}/bin/paster setup-app deployment.ini"
  action :nothing
end

include_recipe "supervisor"

supervisor_service "mediacore" do
  command "uwsgi --ini-paste #{node[:mediacore][:dir]}/deployment.ini --logto #{node[:mediacore][:log_location]}/mediacore.log"
  user node[:mediacore][:user]
  numprocs 1
  autostart true
  autorestart true
  stopsignal "TERM"
  directory node[:mediacore][:dir]
end


attributes/default.rb
# User and Group to run mediacore as
default[:mediacore][:user] = "www-data"
default[:mediacore][:group] = "www-data"

# Basic config options 
default[:mediacore][:smtp_server] = "localhost"
default[:mediacore][:error_email_from] = "mediacore@localhost"
default[:mediacore][:email_to] = "you@yourdomain.com"
default[:mediacore][:host] = "0.0.0.0"
default[:mediacore][:port] = "8080"

# Mediacore default database attributes
default[:mediacore][:db_type] = "postgresql"
default[:mediacore][:db_user] = "mediacore"
default[:mediacore][:db_pass] = "mediacore"
default[:mediacore][:db_address] = "localhost"
default[:mediacore][:database] = "mediacore"
default[:mediacore][:sqlalchemy][:url] = "#{node[:mediacore][:db_type]}://#{node[:mediacore][:db_user]}:#{node[:mediacore][:db_pass]}@#{node[:mediacore][:db_address]}/#{node[:mediacore][:database]}"
default[:mediacore][:sqlalchemy][:echo] = "False" 
default[:mediacore][:sqlalchemy][:pool_recycle] = "3600"

# Basic Options
default[:mediacore][:version] = "0.10.0"
default[:mediacore][:dir] = "/usr/local/mediacore"
default[:mediacore][:data_storage_dir] = "#{node[:mediacore][:dir]}/current/data"
default[:mediacore][:venv] = "#{node[:mediacore][:dir]}/venv"
default[:mediacore][:git_repo] = "git://github.com/mediacore/mediacore-community.git"
default[:mediacore][:log_location] = "/var/log/mediacore"

# More advanced config options
default[:mediacore][:plugins] = [ "*" ] 
default[:mediacore][:uwsgi][:socket] = "/tmp/uwsgi-mediacore.sock"
